rectools/model_selection/cross_validate.py

```diff
@@ -132,34 +86,23 @@ def cross_validate(  # pylint: disable=too-many-locals
             ]
         }
     """
-    interactions = dataset.interactions
-
-    split_iterator = splitter.split(interactions, collect_fold_stats=True)
+    split_iterator = splitter.split(dataset.interactions, collect_fold_stats=True)
 
     split_infos = []
     metrics_all = []
 
     for train_ids, test_ids, split_info in split_iterator:
         split_infos.append(split_info)
 
-        # ### Prepare split data
-        interactions_df_train = interactions.df.iloc[train_ids]  # 1x internal
-        # We need to avoid fitting models on sparse matrices with all zero rows/columns =>
-        # => we need to create a fold dataset which contains only hot users and items for current training
-        fold_dataset = _gen_2x_internal_ids_dataset(
-            interactions_df_train, dataset.user_features, dataset.item_features, prefer_warm_inference_over_cold
+        fold_dataset = dataset.filter_on_interactions_df_row_indexes(
+            row_indexes_to_keep=train_ids,
+            keep_external_ids=True,
+            keep_features_for_removed_entities=prefer_warm_inference_over_cold,
         )
-
-        interactions_df_test = interactions.df.iloc[test_ids]  # 1x internal
-        test_users = interactions_df_test[Columns.User].unique()  # 1x internal
-        catalog = interactions_df_train[Columns.Item].unique()  # 1x internal
-
-        if items_to_recommend is not None:
-            item_ids_to_recommend = dataset.item_id_map.convert_to_internal(
-                items_to_recommend, strict=False
-            )  # 1x internal
-        else:
-            item_ids_to_recommend = None
+        interactions_df_test = dataset.get_raw_interactions().iloc[test_ids]
```

Comment:
Seems really ineffective:
1. You can get raw interactions once before the cycle.
2. We invented internal ids to avoid memory errors caused by big external ids, and here this logic fails.

What can be done now: you can use your new method `filter_on_interactions_df_row_indexes` to get test dataset, and then extract raw interactions from it. It makes sense since test part is usually much smaller than dataset. Yo can also add an argument to your method like `keep_features` and set it to `False` here to make it more effective.

What can we do in future: avoid converting anything to external ids and calculate metrics on internal ones.

Reply:
I did the first one.
Can't see any optimization with calling `filter_on_interactions_df_row_indexes` for test interactions. It calls `get_raw_interactions` under the hood

Reply:
First one was the obvious thought but it still not enough.  We can't keep raw interactions in memory.

Reply:
> Can't see any optimization with calling filter_on_interactions_df_row_indexes for test interactions. It calls get_raw_interactions under the hood

Conclusion: we need to rewrite `filter_on_interactions_df_row_indexes`

Reply:
What do you want to do? get internal interactions.df, apply [test_ids] and then convert to external?

Reply:
Fixed